### DISCUSSION TITLE
How to Cross Platform ?

### DISCUSSION THREAD

**Discussion by pauladam94** [jonasgeiler/lua-fenster#9] [Q&A] [answered]
Do you have any pointers to places where it is explained how to make an executable out of this Lua project ?

Why is the fenster module cross platform ? Is it only because C so Lua is cross platform ?

For example I have made some app that I want to show on a website ? Or publish for people to try on different OS. What do you advise to do for such things ?

**Comment by jonasgeiler** [comment, marked as answer]
> Do you have any pointers to places where it is explained how to make an executable out of this Lua project ?

Packaging an app as a cross-platform executable is way outside the scope of this library.
You can find plenty of tools for this across the internet. [Here is a StackOverflow thread talking about some of these tools.](https://stackoverflow.com/questions/194520/creating-standalone-lua-executables)
Although I do not know if any of them will work with this library and I have never tried it myself, so I unfortunately won't be able to help...

> Why is the fenster module cross platform ? Is it only because C so Lua is cross platform ?

What "cross-platform" means for `fenster` is that you can install `fenster` (using `luarocks install fenster`) and then run a Lua file using `fenster` on every platform (namely Linux, macOS and Windows) and it should work the same everywhere.
To make this work, `fenster` has to hook into 3 different graphics libraries for every platform. That's `X11` for Linux, `gdi32` for Windows and `Cocoa` for macOS. Depending on the platform you use it will hook into one of these to display the window and get user input.
That is also the reason this project is written in C, since you can't just hook into those graphics libraries from Lua itself - you need a layer in-between, which is exactly what `fenster` does.

> For example I have made some app that I want to show on a website ? Or publish for people to try on different OS. What do you advise to do for such things ?

If you wrote a cool program using `fenster` and want to show it to friends and family they would have to install `fenster` themselves (by running `luarocks install fenster`) to be able to run it. This will work for any OS they might use.
Unfortunately `fenster` will not support running in web browsers, so you can't put your program on a website...

Let me know if that clarified some things or opened up even more questions 😅

**Comment by pauladam94** [comment]
Ok thanks for the response. You answered my questions.
